qiskit_experiments/calibration_management/calibration_utils.py

```diff
@@ -12,51 +12,171 @@
 
 """Calibration helper functions"""
 
-from typing import List, Set
-from qiskit.pulse import ScheduleBlock, Call
+from typing import Optional, Set, Tuple
+from functools import lru_cache
+import re
+import retworkx as rx
 
+from qiskit.circuit import ParameterExpression, Parameter
+from qiskit.pulse import ScheduleBlock
 
-def used_in_calls(schedule_name: str, schedules: List[ScheduleBlock]) -> Set[str]:
-    """Find the schedules in the given list that call a given schedule by name.
+from qiskit_experiments.exceptions import CalibrationError
+from qiskit_experiments.calibration_management.calibration_key_types import ScheduleKey
+
+
+# The channel indices need to be parameterized following this regex.
+CHANNEL_PATTERN = r"^ch\d+(\.\d+)*(\$\d+){0,1}$"
+CHANNEL_PATTERN_REGEX = re.compile(CHANNEL_PATTERN)
+
+
+def update_schedule_dependency(schedule: ScheduleBlock, dag: rx.PyDiGraph, key: ScheduleKey):
+    """Update a DAG of schedule dependencies.
+
+    Args:
+        schedule: A ScheduleBlock that potentially has references to other schedules
+            that are already present in the dag.
+        dag: A directed acyclic graph that encodes schedule dependencies using references.
+        key: The schedule key which also contains the qubits.
+    """
+    # First, check if the node is already in the DAG.
+    try:
+        # If it already is in the DAG we remove the existing edges and add the new ones later.
+        parent_idx = dag.nodes().index(repr(key))
```

Comment:
Why does this function work in terms of `repr`? A PyDiGraph can take a ScheduleKey as data directly.

Reply:
Agreed.

Reply:
Good point. Somehow I was under the impression that nodes could not be `ScheduleKeys` but this seems to work well: [667cfbb](https://github.com/Qiskit/qiskit-experiments/pull/889/commits/667cfbb692bbfb1e8f449cf88bc7f7f49ce8242a) note that I kept the `repr` in `ScheduleKey` to keep the output nice when printing the DAG.